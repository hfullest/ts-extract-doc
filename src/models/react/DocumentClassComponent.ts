import {
  ClassDeclaration,
  InterfaceDeclaration,
  Node,
  PropertyDeclaration,
  Symbol,
  Type,
  TypeNode,
  ts,
} from 'ts-morph';
import { BaseDocField, DocumentOptions, SymbolOrOtherType } from '../helper';
import { DocumentClass, DocumentObject } from '../normal';
import { JSDocCustomTagEnum } from '../../utils/jsDocTagDefinition';
import { DocumentParser } from '../index';

export class DocumentClassComponent extends BaseDocField {
  props?: DocumentObject['props'];

  methods?: DocumentObject['methods'];

  constructor(symbolOrOther: SymbolOrOtherType, options: DocumentOptions) {
    const { symbol } = BaseDocField.splitSymbolNodeOrType(symbolOrOther);
    options.$parentSymbol ??= symbol;
    options.$rootSymbol ??= options?.$parentSymbol;
    super(symbolOrOther, options);

    this.#assign(symbolOrOther);
  }

  #assign(symbolOrOther: SymbolOrOtherType) {
    const { symbol } = BaseDocField.splitSymbolNodeOrType(symbolOrOther);
    const declarations = symbol?.getDeclarations(); // 可能有interface重载情况
    if (!declarations?.some((it) => DocumentClassComponent.isTarget(it))) return; // 至少有一个为类组件声明则继续，否则跳出
    declarations?.forEach((declaration) => {
      const node = Node.isVariableDeclaration(declaration)
        ? declaration.getFirstAncestorByKind(ts.SyntaxKind.VariableStatement)
        : declaration;
      this.#handleProps(node!);
    });
  }

  #handleProps(node: Node) {
    const classDeclaration =
      node?.asKind(ts.SyntaxKind.ClassDeclaration) ?? node?.asKind(ts.SyntaxKind.ClassExpression);
    let typeOrTypeNode: TypeNode | Type | undefined;

    if (!typeOrTypeNode) {
      // 构造函数提取第一个属性类型
      const constructorDeclaration = (classDeclaration as ClassDeclaration)?.getConstructors?.()[0];
      const propsNode = constructorDeclaration?.getParameters?.()[0];
      typeOrTypeNode = propsNode?.getTypeNode?.();
    }
    if (!typeOrTypeNode) {
      // React.Component<P,S> 泛型第一个参数提取
      const extendsNode = (classDeclaration as ClassDeclaration)?.getExtends();
      typeOrTypeNode = extendsNode?.getTypeArguments?.()[0];
    }
    if (!typeOrTypeNode) {
      // 提取类组件 props 属性类型
      const propsNode = (classDeclaration as ClassDeclaration)?.getProperty('props');
      typeOrTypeNode = propsNode?.getTypeNode();
      if (!typeOrTypeNode) typeOrTypeNode = propsNode?.getInitializer()?.getType();
    }
    if (!typeOrTypeNode) {
      // 提取类组件 static defaultProps 值类型
      const defaultPropsNode = (classDeclaration as ClassDeclaration)
        ?.getStaticProperty('defaultProps')
        ?.asKind(ts.SyntaxKind.PropertyDeclaration);
      typeOrTypeNode = (defaultPropsNode as PropertyDeclaration)?.getInitializer()?.getType();
    }
    if (!typeOrTypeNode) {
      // 提取 interface 重载 props 属性类型
      const interfaceNode = node?.asKind(ts.SyntaxKind.InterfaceDeclaration);
      typeOrTypeNode = (interfaceNode as InterfaceDeclaration)?.getProperty('props')?.getTypeNode(); // 类组件接口重载声明必须为 props 例如： interface Component { props:{a:string} }
    }
    if (!typeOrTypeNode) return;
    const doc = DocumentParser<DocumentObject>(typeOrTypeNode, this.$options);
    this.props = doc?.props;
    this.methods = doc?.methods;
  }

  static isTarget(symbolOrOther: SymbolOrOtherType) {
    const { node: originNode } = BaseDocField.splitSymbolNodeOrType(symbolOrOther);
    const node = BaseDocField.getCompatAncestorNode(originNode?.getSymbol());
    if (!DocumentClass.isTarget(node!)) return false;
    const jsDocTags = node?.getJsDocs()?.at(-1)?.getTags();
    const isJsxElement = !!jsDocTags?.find((tag) => tag.getTagName() === JSDocCustomTagEnum.reactComponent);
    if (isJsxElement) return true; // 如果手动指定了标签注释，则直接跳过检查，当作组件进行处理
    const extendsNode = node?.getExtends();
    const expressionText = extendsNode?.getExpression?.()?.getText();
    if (!/^(React\.)?(Pure)?Component$/.test(expressionText ?? '')) return false; // 类必须继承自 React.Component、React.PureComponent
    {
      // 校验render函数
      const renderMethods = node?.getMethod?.('render');
      const returnTypeNode = renderMethods?.getReturnTypeNode(); // 如果指定了函数返回类型
      const returnType = returnTypeNode?.getType();
      const returnSubstitionType = renderMethods?.getType()?.getCallSignatures()[0]?.getReturnType();
      const returnSubstitionSymbol = returnSubstitionType?.getSymbol();
      if (
        /^(React\.)?((\w*?)Element|ReactNode)\b/.test(returnType?.getText() ?? '') ||
        returnSubstitionSymbol?.getName() === 'Element'
      ) {
        return true; // 支持 React.(匹配类型)Element、JSX.Element、React.ReactNode
      }
    }
  }
}
