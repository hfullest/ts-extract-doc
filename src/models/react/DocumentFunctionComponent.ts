import { FunctionDeclaration, VariableStatement, ts } from 'ts-morph';
import { DocumentFunction } from '../normal/DocumentFunction';
import { JSDocCustomTagEnum } from '../../utils/jsDocTagDefinition';
import { BaseDocField, DocumentOptions, SymbolOrOtherType } from '../helper';
import { DocumentObject } from '../normal';
import { DocumentParser } from '../index';

// @ts-ignore
export class DocumentFunctionComponent extends DocumentFunction {
  props?: DocumentObject['props'];

  methods?: DocumentObject['methods'];

  constructor(symbolOrOther: SymbolOrOtherType, options: DocumentOptions) {
    const { symbol } = BaseDocField.splitSymbolNodeOrType(symbolOrOther);
    options.$parentSymbol ??= symbol;
    options.$rootSymbol ??= options?.$parentSymbol;
    super(symbolOrOther, options);

    this.#assign(symbolOrOther);
  }

  #assign(symbolOrOther: SymbolOrOtherType) {
    const { symbol } = BaseDocField.splitSymbolNodeOrType(symbolOrOther);
    const functionTypeNode = DocumentFunction.getFunctionTypeNodeBySymbol(symbol!);
    const propsNode = (functionTypeNode as FunctionDeclaration)?.getParameters()?.[0];
    const typeNode = propsNode?.getTypeNode?.();
    const doc = DocumentParser<DocumentObject>(typeNode!, { ...this.$options, $parent: this });
    this.props = doc?.props;
    this.methods = doc?.methods;
  }

  static [Symbol.hasInstance] = (instance: any) => Object.getPrototypeOf(instance).constructor === this;

  static isTarget(nodeOrOther: SymbolOrOtherType) {
    const { node } = BaseDocField.splitSymbolNodeOrType(nodeOrOther);
    const parentNode = DocumentFunction.getCompatAncestorNode<VariableStatement>(node?.getSymbol());
    {
      // TODO: 支持 React.forwardRef 、React.memo
    }
    const functionTypeNode = DocumentFunction.getFunctionTypeNodeBySymbol(node?.getSymbol()!);
    if (!DocumentFunction.isTarget(functionTypeNode!)) return false;
    {
      // 如果手动指定了标签注释，则直接跳过检查，当作组件进行处理
      const jsDocTags = parentNode?.getJsDocs()?.at(-1)?.getTags();
      const isJsxElement = !!jsDocTags?.find((tag) => tag.getTagName() === JSDocCustomTagEnum.reactComponent);
      if (isJsxElement) return true;
    }
    {
      // 如果指定了字面量函数类型，则优先使用类型进行判断 例如：const Button:React.FC = ()=><div></div>
      const variableDeclaration = node?.asKind(ts.SyntaxKind.VariableDeclaration);
      const identifierType = variableDeclaration?.getType();
      if (/^(React\.)?(FC|(\w*?)Component)\b/.test(identifierType?.getText() ?? '')) {
        return true;
      }
    }
    {
      // 支持 React.(匹配类型)Element、JSX.Element、React.ReactNode
      const returnTypeNode = functionTypeNode?.getReturnTypeNode(); // 如果指定了函数返回类型
      const returnType = returnTypeNode?.getType();
      const returnSubstitionType = functionTypeNode?.getType()?.getCallSignatures()[0]?.getReturnType();
      const returnSubstitionSymbol = returnSubstitionType?.getSymbol();
      if (
        /^(React\.)?((\w*?)Element|ReactNode)\b/.test(returnType?.getText() ?? '') ||
        returnSubstitionSymbol?.getName() === 'Element'
      ) {
        return true;
      }
    }

    return false;
  }
}

export default DocumentFunctionComponent;
