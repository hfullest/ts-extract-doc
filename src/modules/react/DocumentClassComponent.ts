import { Node, Symbol } from 'ts-morph';
import { BaseDocField } from '../helper';
import { DocumentFunction } from '../normal/DocumentFunction';

export class DocumentClassComponent extends BaseDocField {
  constructor(symbol: Symbol, parentSymbol: Symbol = symbol, rootSymbol: Symbol = parentSymbol) {
    super(symbol, parentSymbol, rootSymbol);

    this.#assign(symbol);
  }

  #assign(symbol: Symbol) {}

  static isTarget(node: Node) {
    node = BaseDocField.getCompatAncestorNode(node.getSymbol());
    if (!DocumentFunction.isTarget(node)) return false;
  }
}
