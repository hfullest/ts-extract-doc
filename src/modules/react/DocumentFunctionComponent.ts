import { Node, Symbol, VariableStatement, ts } from 'ts-morph';
import BaseDocField from '../BaseDocField';
import DocumentFunction from '../DocumentFunction';
import { JSDocCustomTagEnum } from '../../utils/constants';

export default class DocumentFunctionComponent extends BaseDocField {
  constructor(symbol: Symbol, parentSymbol: Symbol = symbol, rootSymbol: Symbol = parentSymbol) {
    super(symbol, parentSymbol, rootSymbol);

    this.#assign(symbol);
  }

  #assign(symbol: Symbol) {}

  static isTarget(node: Node) {
    debugger;
    const parentNode = DocumentFunction.getCompatAncestorNode<VariableStatement>(node?.getSymbol());
    const functionTypeNode = DocumentFunction.getFunctionTypeNodeBySymbol(node?.getSymbol());
    if (!DocumentFunction.isTarget(functionTypeNode)) return false;
    const jsDocTags = parentNode?.getJsDocs()?.at(-1)?.getTags();
    const isJsxElement = !!jsDocTags?.find((tag) => tag.getTagName() === JSDocCustomTagEnum.reactComponent);
    if (isJsxElement) return true; // 如果手动指定了标签注释，则直接跳过检查，当作组件进行处理
    const variableDeclaration = node?.asKind(ts.SyntaxKind.VariableDeclaration);
    const identifierType = variableDeclaration?.getType();
    if (/^(React\.)?(FC|(\w+?)Component)\b/.test(identifierType?.getText())) {
      return true; // 如果指定了字面量函数类型，则优先使用类型进行判断 例如：const Button:React.FC = ()=><div></div>
    }
    const returnTypeNode = functionTypeNode?.getReturnTypeNode(); // 如果指定了函数返回类型
    const returnType = returnTypeNode?.getType();
    const returnSubstitionType = functionTypeNode
      ?.getType()
      ?.getCallSignatures()[0]
      ?.compilerSignature?.getReturnType();
    const returnSubstitionSymbol = returnSubstitionType?.getSymbol();
    if (/^(React\.)?(\w+?)Element\b/.test(returnType?.getText()) || returnSubstitionSymbol?.getName() === 'Element') {
      return true; // 支持 React.(匹配类型)Element、JSX.Element
    }
    return false;
  }
}
