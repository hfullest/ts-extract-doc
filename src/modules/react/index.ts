export * from './DocumentClassComponent';
export * from './DocumentFunctionComponent';
