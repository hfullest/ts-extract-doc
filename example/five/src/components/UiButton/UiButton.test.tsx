import React from "react";

import { render } from "@testing-library/react";

import { UI_PREFIX, UiButton } from ".";

const prefixCls = `${UI_PREFIX}-btn`;
const antPrefixCls = `ant-btn`;
const role = `test-ui-button`;

describe("【UiButton】case:", () => {
  describe("loading 属性", () => {
    describe("loading: true", () => {
      test(`${prefixCls}-loading 类名添加正常`, () => {
        const { queryByRole } = render(
          <UiButton role={role} loading={true}>
            loading按钮
          </UiButton>
        );
        const btnEle = queryByRole(role);
        expect(btnEle?.className).toContain(`${prefixCls}-loading`);
        expect(btnEle?.className).toContain(`${antPrefixCls}-loading`);
      });
      test("按钮不可点击", async () => {
        const callback = jest.fn();
        const { getByRole } = render(
          <UiButton role={role} loading={true} onClick={callback}>
            loading按钮
          </UiButton>
        );
        getByRole(role).click();
        expect(callback).not.toBeCalled(); // loading 状态按钮不可点击
      });
      test("快照相同", () => {
        const { queryByRole } = render(
          <UiButton role={role} loading={true}>
            loading按钮
          </UiButton>
        );
        expect(queryByRole(role)).toMatchSnapshot();
      });
    });

    describe("loading: {type: 'contain'}", () => {
      test("loading 图标和按钮内容共存", () => {
        const { queryByRole, queryByText } = render(
          <UiButton role={role} loading={{ type: "contain" }}>
            loading按钮
          </UiButton>
        );
        const btnEle = queryByRole(role);
        expect(btnEle?.className).toContain(`${prefixCls}-loading`);
        expect(btnEle?.className).toContain(`${prefixCls}-loading-contain`);
        expect(btnEle?.className).toContain(`${antPrefixCls}-loading`);
        expect(queryByRole("img")).toBeInTheDocument();
        expect(queryByText("loading按钮")).toBeInTheDocument();
        expect(btnEle).toMatchSnapshot();
      });
      test("按钮不可点击", async () => {
        const callback = jest.fn();
        const { getByRole } = render(
          <UiButton
            role={role}
            loading={{ type: "contain" }}
            onClick={callback}
          >
            loading按钮
          </UiButton>
        );
        getByRole(role).click();
        expect(callback).not.toBeCalled(); // loading 状态按钮不可点击
      });
    });

    describe("loading: {icon:<div>loading icon</div>}", () => {
      test("替换loading图标", () => {
        const { queryByRole, queryByText } = render(
          <UiButton role={role} loading={{ icon: <div>loading icon</div> }}>
            loading按钮
          </UiButton>
        );
        expect(queryByRole(role)?.className).not.toContain(
          `${antPrefixCls}-loading`
        );
        expect(queryByRole(role)?.className).toContain(`${prefixCls}-loading`);
        expect(queryByText("loading icon")).toBeInTheDocument();
        expect(queryByRole("img")).not.toBeInTheDocument(); // antd 默认 loading 图标不渲染
        expect(queryByRole(role)).toMatchSnapshot();
      });
    });

    describe("loading: {loading: false}", () => {
      test("不进行 loading", () => {
        const { queryByRole } = render(
          <UiButton role={role} loading={{ loading: false }}>
            loading按钮
          </UiButton>
        );
        expect(queryByRole(role)).not.toContain(`${antPrefixCls}-loading`);
        expect(queryByRole(role)).not.toContain(`${prefixCls}-loading`);
        expect(queryByRole("img")).not.toBeInTheDocument();
        expect(queryByRole(role)).toMatchSnapshot();
      });
    });
  });

  //TODO: disabled 单测
  describe.skip("disabled 属性", () => {
    test("disabled 属性为 boolean 是否工作正常", () => {});

    test("disabled 属性为配置对象是否工作正常", () => {});
  });

  describe.skip("align 属性", () => {});

  describe.skip("tooltip 属性", () => {});
});
