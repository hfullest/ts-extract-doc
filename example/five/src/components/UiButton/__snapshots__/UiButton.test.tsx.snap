// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`【UiButton】case: loading 属性 loading: {icon:<div>loading icon</div>} 替换loading图标 1`] = `
<button
  aria-label="button"
  class="ant-btn ui-btn twoCharNotSpace ui-btn-align-center ui-btn-loading ui-btn-loading-cover"
  role="test-ui-button"
  type="button"
>
  <span
    class="ui-btn-loading-icon"
  >
    <div>
      loading icon
    </div>
  </span>
  <span>
    loading按钮
  </span>
</button>
`;

exports[`【UiButton】case: loading 属性 loading: {loading: false} 不进行 loading 1`] = `
<button
  aria-label="button"
  class="ant-btn ui-btn twoCharNotSpace ui-btn-align-center ui-btn-loading ui-btn-loading-cover"
  role="test-ui-button"
  type="button"
>
  <span>
    loading按钮
  </span>
</button>
`;

exports[`【UiButton】case: loading 属性 loading: {type: 'contain'} loading 图标和按钮内容共存 1`] = `
<button
  aria-label="button"
  class="ant-btn ant-btn-loading ui-btn twoCharNotSpace ui-btn-align-center ui-btn-loading ui-btn-loading-contain"
  role="test-ui-button"
  type="button"
>
  <span
    class="ant-btn-loading-icon"
  >
    <span
      aria-label="loading"
      class="anticon anticon-loading anticon-spin"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="loading"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
  </span>
  <span>
    loading按钮
  </span>
</button>
`;

exports[`【UiButton】case: loading 属性 loading: true 快照相同 1`] = `
<button
  aria-label="button"
  class="ant-btn ant-btn-loading ui-btn twoCharNotSpace ui-btn-align-center ui-btn-loading ui-btn-loading-cover"
  role="test-ui-button"
  type="button"
>
  <span
    class="ant-btn-loading-icon"
  >
    <span
      aria-label="loading"
      class="anticon anticon-loading anticon-spin"
      role="img"
    >
      <svg
        aria-hidden="true"
        data-icon="loading"
        fill="currentColor"
        focusable="false"
        height="1em"
        viewBox="0 0 1024 1024"
        width="1em"
      >
        <path
          d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
        />
      </svg>
    </span>
  </span>
  <span>
    loading按钮
  </span>
</button>
`;
